hummingbot/connector/exchange/bittrex/bittrex_api_order_book_data_source.py

```diff
@@ -1,244 +1,186 @@
-
-        return output
+        except asyncio.CancelledError:
+            raise
+        except Exception:
+            self.logger().error(
+                "Failed to subscribe to public order book and trade channels...",
+                exc_info=True
+            )
+            raise
 
-    async def listen_for_order_book_snapshots(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
-        # Technically this does not listen for snapshot, Instead it periodically queries for snapshots.
-        while True:
+    async def _process_websocket_messages(self, websocket_assistant):
+        """
+        Overriding method for Bittrex signalr
+        """
+        async for raw_message in _socket_stream(conn=websocket_assistant):
             try:
-                async with aiohttp.ClientSession() as client:
-                    for trading_pair in self._trading_pairs:
-                        try:
-                            snapshot: Dict[str, Any] = await self.get_snapshot(client, trading_pair)
-                            snapshot_timestamp: float = time.time()
-                            snapshot_msg: OrderBookMessage = BittrexOrderBook.snapshot_message_from_exchange(
-                                snapshot,
-                                snapshot_timestamp,
-                                metadata={"marketSymbol": trading_pair}
-                            )
-                            output.put_nowait(snapshot_msg)
-                            self.logger().info(f"Saved {trading_pair} snapshots.")
-                            await asyncio.sleep(5.0)
-                        except asyncio.CancelledError:
-                            raise
-                        except Exception:
-                            self.logger().error("Unexpected error.", exc_info=True)
-                            await asyncio.sleep(5.0)
-                    # Waits for delta amount of time before getting new snapshots
-                    this_hour: pd.Timestamp = pd.Timestamp.utcnow().replace(minute=0, second=0, microsecond=0)
-                    next_hour: pd.Timestamp = this_hour + pd.Timedelta(hours=1)
-                    delta: float = next_hour.timestamp() - time.time()
-                    await asyncio.sleep(delta)
+                decoded_msg: Dict[str, Any] = _transform_raw_message(raw_message)
+                if decoded_msg["type"] is None:
+                    continue
+                channel: str = self._channel_originating_message(event_message=decoded_msg)
+                if channel in [self._diff_messages_queue_key, self._trade_messages_queue_key]:
+                    self._message_queue[channel].put_nowait(decoded_msg["results"])
             except Exception:
-                self.logger().error("Unexpected error occurred invoking queryExchangeState", exc_info=True)
-                await asyncio.sleep(5.0)
+                self.logger().exception("Unexpected error while decoding websocket message...")
```

Comment:
We should not be breaking signalr_io encapsulation. To process messages we should be using the events functionality that the library provides (the _process_websocket_messages would have an empty implementation in that case).
Also when working with events we should add extra logic to detect when the connection fails to reconnect it (subscribing to the channels again).
There is already an example of a similar situation with a connector that uses socketio (check CoinDCX connector in the private repository)

Reply:
Reworked the implementation of signalr with inspiration from the CoinDCX implementation.